### ISSUE TITLE
The pre hooks are run synchronously preventing the use of async APIs within the hooks 

### ISSUE THREAD

**Issue by bvanderlaan** [Flux159/cassie-odm#10]
Where the `pre` hooks are executed the library simply iterates over the collection of pre-save hooks and invokes each function. This is done synchronously; there is no use of a `next` callback or `Promise` chaining.

https://github.com/Flux159/cassie-odm/blob/master/lib/model.js#L624

This means that unless I'm missing something, which is always an option, it would be impossible to write pre-hooks which look like this:

```
schema.pre('save', function(model) {
  return doSomethingAsync()
    .then((newValue) => {
       model.hello = newValue;
      });
});
```

Per how this hook function will be invoked the save operation will be send to Cassandra before the `doSomethingAsync()` function returns.

It would be nice if the `pre` hook feature supported either a `next` callback or Promise based pattern to allow for async hooks.

**Comment by Flux159** [comment]
I don't think that any hooks were implemented with async functions in mind (this library is pretty old / existed before node 8.x, so the await keyword wasn't even supported w/out transpiling).

If you want to submit a pull request to support async hooks, I think that the code would be in model.js

https://github.com/Flux159/cassie-odm/blob/f60c6aebeda63413e2d9b82c51c87ee964fd56e7/lib/model.js#L623

The biggest problem is that both the save and remove functions are callback based and not promisified. I don't know when I would have time to implement this as well as test / update the docs (this would require a new major version since it would break backwards compatibility w/ older versions of node if using async/await - which would be preferred).
